### ISSUE TITLE
Ray workers unable to register when used with "venv"-created virtual environment on Windows with Python 3.7.3+

### ISSUE THREAD

**Issue by nirvana-msu** [ray-project/ray#13794]
### What is the problem?

Ray workers are unable to register, causing `ray.init` to hang forever, when used with built-in `venv`-created virtual environment on Windows with Python 3.7.3+

### Reproduction

1. Install Python 3.7.3+  (3.8 and later are all affected as well)
2. Create virtual environment using built-in `venv` (in my case I just used PyCharm). Note that environments created using `virtualenv` are not affected.
3. Install Ray and run `import ray; ray.init(num_cpus=5)`

The main clue as to why `ray.init` hangs can be found within `raylet.out` log file. Say, if we start Ray with 5 workers we'll see the following in the log:
```
[2021-01-29 16:24:53,703 W 12780 6432] worker_pool.cc:471: Received a register request from an unknown worker 13356
[2021-01-29 16:24:53,710 W 12780 6432] worker_pool.cc:471: Received a register request from an unknown worker 3604
[2021-01-29 16:24:53,728 W 12780 6432] worker_pool.cc:471: Received a register request from an unknown worker 17228
[2021-01-29 16:24:53,733 W 12780 6432] worker_pool.cc:471: Received a register request from an unknown worker 13076
[2021-01-29 16:24:53,744 W 12780 6432] worker_pool.cc:471: Received a register request from an unknown worker 18208
[2021-01-29 16:25:23,028 I 12780 6432] worker_pool.cc:375: Some workers of the worker process(8012) have not registered to raylet within timeout.
[2021-01-29 16:25:23,036 I 12780 6432] worker_pool.cc:375: Some workers of the worker process(12484) have not registered to raylet within timeout.
[2021-01-29 16:25:23,045 I 12780 6432] worker_pool.cc:375: Some workers of the worker process(6292) have not registered to raylet within timeout.
[2021-01-29 16:25:23,053 I 12780 6432] worker_pool.cc:375: Some workers of the worker process(5660) have not registered to raylet within timeout.
[2021-01-29 16:25:23,070 I 12780 6432] worker_pool.cc:375: Some workers of the worker process(13064) have not registered to raylet within timeout.
```
i.e. the workers are started and attempt to register promptly, but their PIDs are not recognized by the parent process. That gives us a clue that somehow, process that registers is not the one the parent actually created.

Let's take a step back and just run Python from our venv and check Task Manager for the process(es):
```bash
./venv/Scripts/python -c "import time; time.sleep(30)"
```
![image](https://user-images.githubusercontent.com/5787387/106304845-cb723800-6253-11eb-8f5b-8bcfdde9c4dd.png)

To our surprise, there are actually two processes running! Wtf? Quick googling leads to [this SO post](https://stackoverflow.com/questions/61290972/popen-creates-one-additional-process-in-venv) with helpful comments, and the root cause is discussed in detail in [Issue38905](https://bugs.python.org/issue38905).

In a nutshell, starting from 3.7.3+ built-in `venv` behavior changed such that it always creates two Python processes, first being a simple redirector that spawns the base interpreter. Same happens when Ray launches child Python processes (workers) - pid it gets is that of a redirector, not the actual worker, causing worker registration to fail. 

The issue only affects virtual environments created with built-in `venv`, but not the ones created with `virtualenv`, and only for Python 3.7.3+  (including 3.8 and 3.9, I believe).

It also affects virtual environments created by PyCharm for 3.7.3+ since it's using built-in `venv`.

What confused me the most was that I had an old project using Poetry which was affected by this as well. I then installed Poetry on a new machine and everything worked fine... turns out, Poetry [switched](https://github.com/python-poetry/poetry/pull/2666) from using `venv` to `virtualenv` on Jul 2020, so new Poetry users are not affected. Or at least - not until `virtualenv` is changed to use `venv` internally and this issue suddenly re-surfaces again...

A workaround for anyone affected by this is to not use `venv` but rather use `virtualenv` for creating virtual environment. If you use Poetry, you may need to upgrade Poetry and re-created your environment.

While this `venv` behavior change definitely seems like a regression to me, it seems like it's here to stay - so would be great if Ray handled this. You're going to need to pipe PID from a child process, or something along those lines...

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

P.S. Symptoms of this issue were first described in #12481, but I decided to raise a new issue with a clear description and title.

ping @richardliaw

**Comment by DanielAtKrypton** [comment]
Awesome @nirvana-msu . Have you tried with Python 3.8 already?

**Comment by nirvana-msu** [comment]
@DanielAtKrypton yes, I was using 3.8.6.  This was a permanent change in 3.7.3+ (as described in the linked Python bug report) - so every future version would be affected as well.

**Comment by benlevitas** [comment]
Thank you for this descriptive explanation.
I'm experiencing this same issue when using both venv and virtualenv in python 3.8.7

**Comment by ianormy** [comment]
> Thank you for this descriptive explanation.
> I'm experiencing this same issue when using both venv and virtualenv in python 3.8.7

Like @benlevitas  I also experienced this problem with python 3.8.7 venv and virtualenv. The only way I was able to solve it was to install Anaconda and create a conda virtual environment. That wanted to install python 3.9 (which doesn't work with ray at the moment) so I downgraded it to python 3.8. Once I'd done that and installed everything, it worked fine.

**Comment by burakbagdatli** [comment]
I'm noticing the same problem. Executing ray.init() using a virtual environment makes the environment hang. I cannot even ctrl+c out of it. Killing raylet.exe from task manager is the only way to continue execution.

I don't know if anyone else has tried this but installing ray on the regular python installation works without venv.

**Comment by Stefan-1313** [comment]
I can confirm.

On Windows, I have `Python 3.8.5` and a `venv` and installed ray using `pip install ray`.
When executing `ray.init()` it hangs (ctrl+c) is not working, only closing the console works. It prints the message showing the dashboard-address though.

Then I installed `miniconda` and updated to `Python 3.8.8` and created a conda environment. Then installed ray again using `pip install ray`. Now executing `ray.init()` works as expected.

I would love to have it solved since we normally work in `venv`.

**Comment by simonschoe** [comment]
Push, same issue here, nothing to add to my predecessors ;)

**Comment by aptlin** [comment]
Any updates?

**Comment by DanielAtKrypton** [comment]
Duplicate of https://github.com/ray-project/ray/issues/12481

**Comment by Nesqulck** [comment]
Same Issue on Win10, Ray > 1.4.0, python 3.9 with pipenv which uses venv. Using same packages with global python installation works fine. This further confirms that the issue resolves around venv

**Comment by Allgoerithm** [comment]
I ran into the same issue as @Nesqulck. If this can't be solved very quickly, the installation documentation must be updated to state which environment management tools can be used with Ray, especially as the conflict with venv causes a situation in which it is not easy to detect the root cause of the problem.

**Comment by pcmoritz** [comment]
We think this is the same issue as https://github.com/ray-project/ray/issues/18951, it should be fixed when https://github.com/ray-project/ray/pull/19014 is merged :)

**Comment by pcmoritz** [comment]
I think this issue has now been fully fixed on master and will be available with the 1.10 release! I'm closing this issue but please let us know if it still happens with 1.10.

**Comment by ky1emoon** [comment]
Still hanging error occurs on [Ubuntu 20.04, Python 3.8, Ray 1.10]

It works on the base interpreter,

but using venv environment still occurs hanging error on ray.init()

Note that virtualenv environment dose not occur this hanging problem.

**Comment by nrclaudio** [comment]
This still happens. 

ray.init() hangs with a clean conda installation `conda create -n ray python=3.8; pip install ray`

[CentOS Stream release 8. python=3.8, ray=1.11.0]

`cat session_latest/logs/raylet.err`

`[2022-04-09 01:07:03,753 E 2174263 2174263] (raylet) worker_pool.cc:481: Some workers of the worker process(2174338) have not registered within the timeout. The process is still alive, probably it's hanging during start.`
